src/test/java/io/confluent/idesidecar/restapi/kafkarest/RecordsV3ApiImplIT.java

```diff
@@ -0,0 +1,401 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static io.confluent.idesidecar.restapi.util.ResourceIOUtil.loadResource;
+import static org.hamcrest.Matchers.*;
+import static org.junit.jupiter.api.Assertions.*;
+
+@QuarkusIntegrationTest
+@Tag("io.confluent.common.utils.IntegrationTest")
+@TestProfile(NoAccessFilterProfile.class)
+class RecordsV3ApiImplIT extends ConfluentLocalTestBed {
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+  record RecordData(
+      SchemaManager.SchemaFormat schemaFormat,
+      String rawSchema,
+      Object data
+  ) {
+
+    @Override
+    public String toString() {
+      return "(data = %s, schemaFormat = %s, schema = %s)"
+          .formatted(data, schemaFormat, rawSchema);
+    }
+  }
+
+  static RecordData schemaData(SchemaManager.SchemaFormat format, String jsonData) {
+    var schema = getProductSchema(format);
+
+    try {
+      return new RecordData(format, schema, OBJECT_MAPPER.readTree(jsonData));
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static String getProductSchema(SchemaManager.SchemaFormat format) {
+    return switch (format) {
+      case JSON: {
+        yield loadResource("schemas/product.schema.json").translateEscapes();
+      }
+      case AVRO: {
+        yield loadResource("schemas/product.avsc").translateEscapes();
+      }
+      case PROTOBUF: {
+        yield loadResource("schemas/product.proto").translateEscapes();
+      }
+    };
+  }
+
+  @Nested
+  // Do we really need to specify this TestProfile again? Yes. Why? I don't know.
+  @TestProfile(NoAccessFilterProfile.class)
+  class SadPath {
+    @Test
+    void shouldThrowNotFoundWhenClusterDoesNotExist() {
+      givenConnectionId()
+          .body(createProduceRequest(
+              null, "key", null, "value", null
+          ))
+          .post("/kafka/v3/clusters/non-existent-cluster/topics/foo/records")
+          .then()
+          .statusCode(404)
+          .body("message", equalTo("Kafka cluster 'non-existent-cluster' not found."));
+    }
+
+    @Test
+    void shouldThrowNotFoundWhenTopicDoesNotExist() {
+      produceRecordThen(null, "non-existent-topic", null, null)
+          .statusCode(404)
+          .body("message", equalTo("This server does not host this topic-partition."));
+    }
+
+    @Test
+    void shouldThrowNotFoundWhenPartitionDoesNotExist() {
+      var topic = randomTopicName();
+      createTopic(topic, 3, 1);
+      produceRecordThen(10, topic, "key", "value")
+          .statusCode(404)
+          .body(
+              "message", equalTo("This server does not host this topic-partition."));
+    }
+
+    private static Stream<Arguments> keyAndValueSchemaVersions() {
+      return Stream.of(
+          Arguments.of(40, null),
+          Arguments.of(null, 40),
+          Arguments.of(40, 40)
+      );
+    }
```

Comment:
WDYT about making this non-private and adding a JavaDoc before it that describes what this is producing?
```suggestion
    /**
     * Inputs for test cases for when the key and value schema versions are not found.
     * @return the Arguments for the test cases
     */
    static Stream<Arguments> keyAndValueSchemaVersions() {
      return Stream.of(
          Arguments.of(40, null),
          Arguments.of(null, 40),
          Arguments.of(40, 40)
      );
    }
```